chain/network/src/types.rs

```diff
@@ -206,6 +225,50 @@ impl Handshake {
     }
 }
 
+// Use custom deserializer for HandshakeV2. Try to read version of the other peer from the header.
+// If the version is supported then fallback to standard deserializer.
+impl BorshDeserialize for Handshake {
+    fn deserialize(buf: &mut &[u8]) -> std::io::Result<Self> {
+        // Detect the current and oldest supported version from the header
+        if buf.len() < 4 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+            ));
+        }
+
+        let version = u32::from_le_bytes(buf[..4].try_into().unwrap());
+
+        if OLDEST_BACKWARD_COMPATIBLE_PROTOCOL_VERSION <= version && version <= PROTOCOL_VERSION {
+            // If we support this version, then try to deserialize with custom deserializer
+            match version {
+                _ => HandshakeAutoDes::deserialize(buf).map(Into::into),
```

Comment:
this match doesn't do anything and should just be removed

Reply:
It was as a signal for future versions to use custom deserializers, but probably it won't be used anyway in the current design!